Add a CLI for end-to-end evaluation

perlitz opened the pull request:
Adding an end-to-end evaluation script, keeping as close as possible to the API of to the lm_eval https://github.com/EleutherAI/lm-evaluation-harness.

here is an example command:

```bash
unitxt-eval \
    --tasks "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix" \
    --split validation \
    --model cross_provider \
    --model_args "model_name=llama-3-3-70b-instruct,max_tokens=256" \
    --limit 100 \
    --output_dir ./debug_output/bird_remote \
    --log_samples
```
    
or for a local model:
    
```bash
unitxt-eval \
    --tasks "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix" \
    --split validation \ 
    --model hf \
    --model_args "pretrained=HuggingFaceTB/SmolLM2-135M-Instruct,max_new_tokens=256" \
    --limit 100 \
    --output_dir ./debug_output/bird_remote \
    --log_samples 
```

for example, the results for the first run looks like this:

```json
{
    "environment_info": {
        "timestamp_utc": "2025-04-02T10:10:18.401969Z",
        "command_line_invocation": [
            "/opt/homebrew/Caskroom/miniforge/base/envs/sage/bin/unitxt-eval",
            "--tasks",
            "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix",
            "--split",
            "validation",
            "--model",
            "cross_provider",
            "--model_args",
            "model_name=llama-3-3-70b-instruct,max_tokens=256",
            "--limit",
            "100",
            "--output_dir",
            "./debug_output/bird_remote",
            "--log_samples"
        ],
        "parsed_arguments": {
            "tasks": "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix",
            "split": "validation",
            "limit": 100,
            "model": "cross_provider",
            "model_args": {
                "max_tokens": 256
            },
            "output_dir": "./debug_output/bird_remote",
            "output_file_prefix": "evaluation_results",
            "log_samples": true,
            "verbosity": "INFO",
            "trust_remote_code": false,
            "disable_hf_cache": false,
            "cache_dir": null
        },
        "unitxt_version": "1.21.0",
        "unitxt_commit_hash": "4a76453c309e47e44e9c5043da384d69e1ab0bae",
        "python_version": "3.10.16",
        "system": "Darwin",
        "system_version": "Darwin Kernel Version 24.2.0: Fri Dec  6 19:02:41 PST 2024; root:xnu-11215.61.5~2/RELEASE_ARM64_T6030",
        "installed_packages": {
            "datasets": "3.5.0",
            "pandas": "2.2.3",
            "tomli": "2.0.1",
            "async-timeout": "5.0.1",
            "idna": "3.10",
            "distlib": "0.3.9",
            "diskcache": "5.6.3",
            "codespell": "2.4.1",
            "unitxt": "1.21.0",
            "pytz": "2025.2",
            "importlib_metadata": "8.0.0",
            "zipp": "3.19.2",
            "anyio": "4.9.0",
            "requests": "2.32.3",
            "frozenlist": "1.5.0",
            "attrs": "25.3.0",
            "setuptools": "75.8.2",
            "h11": "0.14.0",
            "networkx": "3.4.2",
            "accelerate": "1.6.0",
            "detect-secrets": "1.5.0",
            "ruff": "0.11.2",
            "yaspin": "3.1.0",
            "rpds-py": "0.24.0",
            "tiktoken": "0.9.0",
            "tqdm": "4.67.1",
            "urllib3": "2.3.0",
            "exceptiongroup": "1.2.2",
            "click": "8.1.8",
            "typing-inspection": "0.4.0",
            "tzdata": "2025.2",
            "pre_commit": "4.2.0",
            "tokenizers": "0.21.1",
            "tenacity": "9.0.0",
            "virtualenv": "20.30.0",
            "jsonschema": "4.23.0",
            "openai": "1.70.0",
            "python-dotenv": "1.1.0",
            "httpretty": "1.1.4",
            "multiprocess": "0.70.16",
            "pydantic_core": "2.33.0",
            "xxhash": "3.5.0",
            "referencing": "0.36.2",
            "func_timeout": "4.3.5",
            "pip": "25.0.1",
            "regex": "2024.11.6",
            "identify": "2.6.9",
            "fsspec": "2024.12.0",
            "sniffio": "1.3.1",
            "nodeenv": "1.9.1",
            "psutil": "7.0.0",
            "jsonschema-specifications": "2024.10.1",
            "cfgv": "3.4.0",
            "python-dateutil": "2.9.0.post0",
            "mpmath": "1.3.0",
            "yarl": "1.18.3",
            "charset-normalizer": "3.4.1",
            "numpy": "2.2.4",
            "packaging": "24.2",
            "sqlglot": "26.12.1",
            "torch": "2.6.0",
            "scipy": "1.15.2",
            "certifi": "2025.1.31",
            "evaluate": "0.4.3",
            "platformdirs": "4.2.2",
            "httpcore": "1.0.7",
            "Jinja2": "3.1.6",
            "wheel": "0.43.0",
            "transformers": "4.50.3",
            "filelock": "3.18.0",
            "pyarrow": "19.0.1",
            "dill": "0.3.8",
            "distro": "1.9.0",
            "aiohappyeyeballs": "2.6.1",
            "fuzzywuzzy": "0.18.0",
            "propcache": "0.3.1",
            "sympy": "1.13.1",
            "aiohttp": "3.11.15",
            "aiosignal": "1.3.2",
            "PyYAML": "6.0.2",
            "httpx": "0.28.1",
            "safetensors": "0.5.3",
            "sqlparse": "0.5.3",
            "litellm": "1.65.1",
            "multidict": "6.3.0",
            "typing_extensions": "4.12.2",
            "termcolor": "2.3.0",
            "pydantic": "2.11.1",
            "annotated-types": "0.7.0",
            "six": "1.17.0",
            "jiter": "0.9.0",
            "huggingface-hub": "0.30.1",
            "tabulate": "0.9.0",
            "MarkupSafe": "3.0.2",
            "Sage": "1.0",
            "jaraco.text": "3.12.1",
            "jaraco.collections": "5.1.0",
            "jaraco.functools": "4.0.1",
            "jaraco.context": "5.3.0",
            "inflect": "7.3.1",
            "autocommand": "2.2.2",
            "backports.tarfile": "1.2.0",
            "typeguard": "4.3.0",
            "more-itertools": "10.3.0"
        }
    },
    "global_scores": {
        "num_of_instances": 100,
        "anls": 0.5449067219688716,
        "score": 0.45,
        "score_name": "non_empty_execution_accuracy",
        "sql_exact_match": 0.05,
        "sqlparse_equivalence": 0.03,
        "sqlparse_validity": 1.0,
        "sqlglot_equivalence": 0.06,
        "sqlglot_validity": 1.0,
        "sqlglot_optimized_equivalence": 0.08,
        "sql_exact_match_ci_low": 0.02,
        "sql_exact_match_ci_high": 0.1,
        "sqlparse_equivalence_ci_low": 0.01,
        "sqlparse_equivalence_ci_high": 0.08,
        "sqlglot_equivalence_ci_low": 0.02,
        "sqlglot_equivalence_ci_high": 0.12,
        "score_ci_low": 0.35,
        "score_ci_high": 0.55,
        "sqlglot_optimized_equivalence_ci_low": 0.04,
        "sqlglot_optimized_equivalence_ci_high": 0.15,
        "execution_accuracy": 0.45,
        "subset_non_empty_execution_result": 0.52,
        "predicted_error": 0.06,
        "predicted_sql_runtime": 0.002272778769256547,
        "gold_sql_runtime": 0.0026918408303754403,
        "pred_to_gold_runtime_ratio": 1.2055306179668457,
        "non_empty_execution_accuracy": 0.45,
        "gold_error": 0.0,
        "non_empty_gold_df": 0.94,
        "execution_accuracy_ci_low": 0.35,
        "execution_accuracy_ci_high": 0.55,
        "subset_non_empty_execution_result_ci_low": 0.42,
        "subset_non_empty_execution_result_ci_high": 0.62,
        "predicted_sql_runtime_ci_low": 0.001964889455919157,
        "predicted_sql_runtime_ci_high": 0.002637987524720579,
        "gold_sql_runtime_ci_low": 0.002245982536213698,
        "gold_sql_runtime_ci_high": 0.003846346007396595,
        "non_empty_execution_accuracy_ci_low": 0.35,
        "non_empty_execution_accuracy_ci_high": 0.55
    }
}
```

elronbandel reviewed (commented):
This is absolutely brilliant. Very useful.

yoavkatz commented on src/unitxt/evaluate_cli.py:
```diff
@@ -0,0 +1,662 @@
+    # --- Local Hugging Face Model (using HFAutoModelInferenceEngine) ---
+    if args.model.lower() == "hf":
+        # Check for transformers and torch only when hf model is selected
+        if (
```
I don't think checking installed libraries is need.  Unitxt inference engine already do this, and it will be double distances.

perlitz replied:
sure, i'll remove these

yoavkatz commented on src/unitxt/evaluate_cli.py:
```diff
@@ -0,0 +1,662 @@
+        )
+
+    # --- Remote Model (CrossProviderInferenceEngine) ---
+    elif args.model.lower() == "generic_remote":
```
Should be called cross_provider to be consistent.

perlitz replied:
Done

perlitz commented:
@yoavkatz @elronbandel @eladven  I've added some metadata to the results files so one can reproduce,
added this to the first comment, see if you have some more suggestions

elronbandel commented on docs/docs/cli.rst:
```diff
@@ -0,0 +1,255 @@
+    .. code-block:: bash
+
+       unitxt-evaluate \
+           --tasks "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix" \
```
```suggestion
           --task "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix" \
```
In the code it is task I think

elronbandel commented on src/unitxt/evaluate_cli.py:
```diff
@@ -0,0 +1,939 @@
+        "-t",
+        type=str,
+        required=True,
+        help="Unitxt task/dataset identifier string.\n"
```
Maybe worth adding link to documentation of the api https://www.unitxt.ai/en/latest/unitxt.standard.html#module-unitxt.standard

elronbandel commented on src/unitxt/evaluate_cli.py:
```diff
@@ -0,0 +1,939 @@
+        )
+
+        # Apply unitxt settings within a context manager
+        with configure_unitxt_settings(args):
```
```suggestion
        with settings.context(**args):
```
Built in capability

elronbandel commented on src/unitxt/evaluate_cli.py:
```diff
@@ -0,0 +1,939 @@
+logger = get_logger()
+
+
+def _parse_key_value_string(value: str) -> Optional[Dict[str, Any]]:
+    """Parses a key=value comma-separated string into a dict."""
+    parsed_dict = {}
+    for item in value.split(","):
+        item = item.strip()
+        if not item:
+            continue
+        parts = item.split("=", 1)
+        if len(parts) == 2:
+            key, val_str = parts
+            key = key.strip()  # Strip whitespace from key
+            # Attempt to convert value to int, float, or bool
+            val: Any
+            try:
+                val = int(val_str)
+            except ValueError:
+                try:
+                    val = float(val_str)
+                except ValueError:
+                    if val_str.lower() == "true":
+                        val = True
+                    elif val_str.lower() == "false":
+                        val = False
+                    else:
+                        val = val_str  # Keep as string
+            parsed_dict[key] = val
+        else:
+            logger.warning(
+                f"Could not parse argument part: '{item}'. Expected format 'key=value'."
+            )
+    return parsed_dict if parsed_dict else None
```
```suggestion
from .parsing_utils import parse_key_equals_value_string_to_dict as _parse_key_value_string
```

elronbandel reviewed (changes requested):
Most important to rely on internal parsing (as it was carefully written and tested:

https://github.com/IBM/unitxt/blob/main/src/unitxt/dataset_utils.py

oktie commented:
I did a test run and it works well for me. On my setup I had to do:
```
export UNITXT_ALLOW_UNVERIFIED_CODE=True
pip install ".[remote_inference,tests]" 
```
and this is what I tested:
```
unitxt-evaluate \
    --task "card=cards.text2sql.bird,template=templates.text2sql.you_are_given_with_hint_with_sql_prefix" \
    --split validation \
    --model cross_provider \
    --model_args "model_name=llama-3-3-70b-instruct,max_tokens=256" \
    --limit 100 \
    --output_path ./debug_output/bird_remote \
    --log_samples
```

Looks very nice!
